www-reputation-credentials-protocol.md

```diff
@@ -0,0 +1,265 @@
+at the reception of a HTLC forward request. A cryposystem satisfying those requirements could be to
+leverage EC [blinded signatures](https://sceweb.sce.uhcl.edu/yang/teaching/csci5234WebSecurityFall2011/Chaum-blind-signatures.PDF).
+
+Additionally, this proposal is extending the `channel_update` message to announce routing fees paid on the
+CLTV duration, therefore allowing compensation of the routing hops for the transport of long-term held
+class of packets (e.g [hold-invoice](https://github.com/lightningnetwork/lnd/pull/2022), [swaps](https://github.com/ElementsProject/peerswap/blob/master/docs/peer-protocol.md)).
+
+# Protocol extensions
+
+This proposal introduces modifications to the following existent BOLT data structures:
+- one new TLV record `credentials_payload` is added to the BOLT4 onion payload
+- one new failure message `routing_policy_error` is added in the list of BOLT4 error messages
+- two new TLV records `fee_base_block_grace_threshold` and `fee_base_block` are added to `channel_update`
+
+Two new data structures are introduced:
+- a BOLT7 gossip message `routing_policy`
+- a credential data format: a 32-byte string of random data
+
+# Protocol Phase
+
+This proposal introduces a completely new dissemination phase unknown from the current LN protocol
+set of operations.
+
+Additionally, this proposal modifies few other protocol operations:
+- the routing & payment path construction
+- the HTLC forward phase (i.e ["Accepting and Forwarding a Payment"](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#accepting-and-forwarding-a-payment))
+- the HTLC settlement phase (i.e ["Removing an HTLC"](https://github.com/lightning/bolts/blob/master/02-peer-protocol.md#removing-an-htlc-update_fulfill_htlc-update_fail_htlc-and-update_fail_malformed_htlc))
+
+## Credentials dissemination phase
+
+The dissemination phase consists of the satisfaction of an acquisition method announced by the
+routing hop by the HTLC sender.
+
+```
+			2.
+			    BOLT-12 offers
+              -------------------------------------------
+             /	   3.				         \
+	 ___V_____    HTLC     _________    HTLC     _____\____
+	|	  |---------->|	        |---------->|	       |
+	|	  |           |         |           |	       |
+	|  Alice  | 1.        |   Bob   |           |  Caroll  |
+	|	  |  gossips  |	        |  gossips  |	       |
+	|_________|<----------|_________|<----------|__________|
+	    ^					   /
+	     \					  /
+	      ------------------------------------
+		       onions(credentials)
+		    4.
+
+```
+
+In this diagram:
+1. Alice discovers Caroll's `routing_policy` gossip.
+2. Alice fetches an offer from Caroll to pay the `credential_to_liquidity_unit` announced.
+3. Alice sends a HTLC to Caroll, the onion payload contains unsigned credentials.
+4. Caroll receives the HTLC, counter-signed the credentials and send them back to Alice by onions.
```

Comment:
Why are these send back via a separate onion message and not in the `update_fulfill_htlc` as described for ongoing signing below?

Reply:
The quantity of signed credentials, and as such the size of the wrapping onion, could be a leak of Alice expected future usage of the payment paths. Therefore, dissociating the backward communication channel from the payment routing hops could improve privacy. The signed credentials could be fan-out in multiple 32k-byte onions.

Reply:
Another option is to make the payload same size for any number of creds somehow... Maybe a static big payload (part of which could be empty), or some fancy compressed data structure with the size independent from the number of creds.

Reply:
I think there is two constraint here, the bandwidth consumption if we use the max onion payload size, whatever the number of credentials delivered. Another constraint, the information efficiency of the credentials. Introducing multiple ranges of credentials could be a correlation traffic vector.